File: chatbi-server/app/chat_service.py
import uuid
from typing import Dict, Optional
from app.models import ChatRequest, ChatResponse, SemanticSQL
from app.semantic_sql_converter import semantic_sql_converter, mysql_sql_generator
from app.database import db_manager
from app.database_admin import database_admin_service
import logging

class ChatService:
    """聊天服务"""
    
    def __init__(self):
        self.conversations: Dict[str, list] = {}
    
    def process_chat_message(self, request: ChatRequest) -> ChatResponse:
        """处理聊天消息"""
        try:
            logger = logging.getLogger("chatbi.service")
            # 获取或创建会话ID
            conversation_id = request.conversation_id or str(uuid.uuid4())
            
            # 获取数据库连接
            selected_connection = None
            if request.database_connection_id:
                selected_connection = database_admin_service.get_connection(request.database_connection_id)
                if not selected_connection:
                    logger.warning("Database connection not found: %s", request.database_connection_id)
                    # 如果指定的连接不存在，使用默认连接
                    selected_connection = database_admin_service.get_active_connection()
                else:
                    logger.info("Using specified database connection: %s (%s)", 
                               selected_connection.name, selected_connection.id)
            else:
                # 使用默认连接
                selected_connection = database_admin_service.get_active_connection()
                logger.info("Using default database connection: %s", 
                           selected_connection.name if selected_connection else "None")
            
            # 初始化会话历史
            if conversation_id not in self.conversations:
                self.conversations[conversation_id] = []
            
            # 添加用户消息到会话历史
            self.conversations[conversation_id].append({
                "role": "user",
                "content": request.message
            })
            
            # 转换自然语言为语义SQL
            logger.info("Converting NL to semantic SQL: cid=%s", conversation_id)
            semantic_sql = semantic_sql_converter.convert_to_semantic_sql(request.message)
            debug_ollama = getattr(semantic_sql_converter, 'last_debug', None)
            
            # 生成MySQL SQL语句
            mysql_sql = mysql_sql_generator.generate_mysql_sql(semantic_sql)
            logger.info("Generated MySQL SQL: cid=%s sql=%s", conversation_id, mysql_sql)
            
            # 生成响应消息
            response_message = self._generate_response_message(
                request.message, 
                semantic_sql, 
                mysql_sql
            )
            
            # 添加助手回复到会话历史
            self.conversations[conversation_id].append({
                "role": "assistant",
                "content": response_message,
                "semantic_sql": semantic_sql.dict(),
                "mysql_sql": mysql_sql
            })
            
            return ChatResponse(
                response=response_message,
                sql_query=mysql_sql,
                semantic_sql=semantic_sql,
                conversation_id=conversation_id,
                debug_ollama=debug_ollama
            )
            
        except Exception as e:
            logging.getLogger("chatbi.service").exception("process_chat_message error: %s", e)
            error_message = f"处理消息时发生错误: {str(e)}"
            # 尝试获取调试信息，即使在异常情况下
            debug_ollama = getattr(semantic_sql_converter, 'last_debug', None)
            return ChatResponse(
                response=error_message,
                conversation_id=request.conversation_id or str(uuid.uuid4()),
                debug_ollama=debug_ollama
            )
    
    def _generate_response_message(
        self, 
        user_message: str, 
        semantic_sql: SemanticSQL, 
        mysql_sql: str
    ) -> str:
        """生成响应消息"""
        
        # 检查是否是查询请求
        if any(keyword in user_message.lower() for keyword in [
            "查询", "显示", "获取", "找出", "统计", "计算", "求和", "平均", "最大", "最小"
        ]):
            
            response = f"我已经将您的查询转换为SQL语句：\n\n"
            response += f"**语义SQL结构：**\n"
            response += f"- 涉及表: {', '.join(semantic_sql.tables) if semantic_sql.tables else '未指定'}\n"
            response += f"- 查询列: {', '.join(semantic_sql.columns) if semantic_sql.columns else '所有列'}\n"
            
            if semantic_sql.conditions:
                response += f"- 筛选条件: {len(semantic_sql.conditions)}个条件\n"
            
            if semantic_sql.aggregations:
                response += f"- 聚合函数: {len(semantic_sql.aggregations)}个\n"
            
            if semantic_sql.joins:
                response += f"- 表连接: {len(semantic_sql.joins)}个\n"
            
            response += f"\n**生成的MySQL SQL：**\n```sql\n{mysql_sql}\n```\n\n"
            response += "您想要执行这个SQL查询吗？"
            
        else:
            response = f"我理解您想要：{user_message}\n\n"
            response += "我已经生成了相应的SQL查询语句。您需要我执行查询并返回结果吗？"
        
        return response
    
    def execute_sql_and_update_response(self, conversation_id: str, sql: str) -> Dict:
        """执行SQL并更新响应"""
        try:
            # 获取当前活动的数据库连接
            active_connection = database_admin_service.get_active_connection()
            
            # 执行SQL查询
            if active_connection:
                execution_result = db_manager.execute_query(sql, active_connection)
            else:
                execution_result = db_manager.execute_query(sql)
            
            # 更新会话历史中的最后一条助手消息
            if (conversation_id in self.conversations and 
                self.conversations[conversation_id] and 
                self.conversations[conversation_id][-1]["role"] == "assistant"):
                
                self.conversations[conversation_id][-1]["execution_result"] = execution_result.dict()
            
            return execution_result.dict()
            
        except Exception as e:
            return {
                "success": False,
                "error": str(e),
                "row_count": 0
            }
    
    def get_conversation_history(self, conversation_id: str) -> list:
        """获取会话历史"""
        return self.conversations.get(conversation_id, [])
    
    def clear_conversation(self, conversation_id: str):
        """清除会话历史"""
        if conversation_id in self.conversations:
            del self.conversations[conversation_id]

# 创建全局实例
chat_service = ChatService()

File: chatbi-server-java/src/main/java/com/chatbi/service/ChatService.java
package com.chatbi.service;

import com.chatbi.model.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;

@Service
public class ChatService {
    private static final Logger logger = LoggerFactory.getLogger(ChatService.class);
    
    @Autowired
    private SemanticSQLConverter semanticSQLConverter;
    
    @Autowired
    private MySQLSQLGenerator mysqlSQLGenerator;
    
    @Autowired
    private DatabaseManager databaseManager;
    
    @Autowired
    private DatabaseConnectionService databaseConnectionService;
    
    private final Map<String, List<Map<String, Object>>> conversations = new ConcurrentHashMap<>();

    public ChatResponse processChatMessage(ChatRequest request) {
        try {
            logger.info("Incoming chat: conversation_id={}, message={}", 
                request.getConversationId(), request.getMessage());
            
            // Get or create conversation ID
            String conversationId = request.getConversationId();
            if (conversationId == null || conversationId.trim().isEmpty()) {
                conversationId = UUID.randomUUID().toString();
            }
            
            // Get database connection
            DatabaseConnection selectedConnection = getSelectedConnection(request.getDatabaseConnectionId());
            
            // Initialize conversation history
            if (!conversations.containsKey(conversationId)) {
                conversations.put(conversationId, new ArrayList<>());
            }
            
            // Add user message to conversation history
            Map<String, Object> userMessage = new HashMap<>();
            userMessage.put("role", "user");
            userMessage.put("content", request.getMessage());
            conversations.get(conversationId).add(userMessage);
            
            // Convert natural language to semantic SQL (with short context: include previous one user input if exists)
            logger.info("Converting NL to semantic SQL: cid={}", conversationId);
            String shortContextInput = request.getMessage();
            List<Map<String, Object>> history = conversations.get(conversationId);
            if (history != null && history.size() >= 2) {
                // Find the most recent previous user message before the current one
                for (int i = history.size() - 2; i >= 0; i--) {
                    Object role = history.get(i).get("role");
                    if (role != null && "user".equals(role.toString())) {
                        Object prevContent = history.get(i).get("content");
                        if (prevContent != null) {
                            shortContextInput = "上一次用户输入（供参考）：" + prevContent.toString() + "\n当前用户输入：" + request.getMessage();
                        }
                        break;
                    }
                }
            }
            SemanticSQL semanticSQL = semanticSQLConverter.convertToSemanticSQL(shortContextInput, selectedConnection);
            Map<String, Object> debugOllama = semanticSQLConverter.getLastDebug();
            
            // Generate MySQL SQL statement
            String mysqlSQL = mysqlSQLGenerator.generateMySQLSQL(semanticSQL);
            logger.info("Generated MySQL SQL: cid={} sql={}", conversationId, mysqlSQL);
            
            // Generate response message
            String responseMessage = generateResponseMessage(request.getMessage(), semanticSQL, mysqlSQL);
            
            // Add assistant reply to conversation history
            Map<String, Object> assistantMessage = new HashMap<>();
            assistantMessage.put("role", "assistant");
            assistantMessage.put("content", responseMessage);
            assistantMessage.put("semantic_sql", semanticSQL);
            assistantMessage.put("mysql_sql", mysqlSQL);
            conversations.get(conversationId).add(assistantMessage);
            
            return new ChatResponse(
                responseMessage,
                mysqlSQL,
                semanticSQL,
                conversationId,
                null,
                debugOllama
            );
            
        } catch (Exception e) {
            logger.error("process_chat_message error: {}", e.getMessage(), e);
            String errorMessage = "处理消息时发生错误: " + e.getMessage();
            Map<String, Object> debugOllama = semanticSQLConverter.getLastDebug();
            
            return new ChatResponse(
                errorMessage,
                null,
                null,
                request.getConversationId() != null ? request.getConversationId() : UUID.randomUUID().toString(),
                null,
                debugOllama
            );
        }
    }

    public Map<String, Object> executeSqlAndUpdateResponse(String conversationId, String sql) {
        try {
            // Get current active database connection
            DatabaseConnection activeConnection = databaseConnectionService.getActiveConnection();
            
            // Execute SQL query
            SQLExecutionResponse executionResult;
            if (activeConnection != null) {
                executionResult = databaseManager.executeQuery(sql, activeConnection);
            } else {
                executionResult = databaseManager.executeQuery(sql);
            }
            
            // Update the last assistant message in conversation history
            if (conversations.containsKey(conversationId) && 
                !conversations.get(conversationId).isEmpty()) {
                
                List<Map<String, Object>> history = conversations.get(conversationId);
                Map<String, Object> lastMessage = history.get(history.size() - 1);
                
                if ("assistant".equals(lastMessage.get("role"))) {
                    lastMessage.put("execution_result", executionResult);
                }
            }
            
            return Map.of(
                "success", executionResult.isSuccess(),
                "data", executionResult.getData(),
                "error", executionResult.getError(),
                "row_count", executionResult.getRowCount()
            );
            
        } catch (Exception e) {
            logger.error("Error executing SQL and updating response: {}", e.getMessage(), e);
            return Map.of(
                "success", false,
                "error", e.getMessage(),
                "row_count", 0
            );
        }
    }

    public List<Map<String, Object>> getConversationHistory(String conversationId) {
        return conversations.getOrDefault(conversationId, new ArrayList<>());
    }

    public void clearConversation(String conversationId) {
        conversations.remove(conversationId);
    }

    private DatabaseConnection getSelectedConnection(String databaseConnectionId) {
        if (databaseConnectionId != null && !databaseConnectionId.trim().isEmpty()) {
            Optional<DatabaseConnection> connection = databaseConnectionService.getConnection(databaseConnectionId);
            if (connection.isPresent()) {
                logger.info("Using specified database connection: {} ({})", 
                    connection.get().getName(), connection.get().getId());
                return connection.get();
            } else {
                logger.warn("Database connection not found: {}", databaseConnectionId);
            }
        }
        
        // Use default connection
        DatabaseConnection activeConnection = databaseConnectionService.getActiveConnection();
        logger.info("Using default database connection: {}", 
            activeConnection != null ? activeConnection.getName() : "None");
        return activeConnection;
    }

    private String generateResponseMessage(String userMessage, SemanticSQL semanticSQL, String mysqlSQL) {
        // Check if it's a query request
        String lowerMessage = userMessage.toLowerCase();
        if (lowerMessage.contains("查询") || lowerMessage.contains("显示") || 
            lowerMessage.contains("获取") || lowerMessage.contains("找出") || 
            lowerMessage.contains("统计") || lowerMessage.contains("计算") || 
            lowerMessage.contains("求和") || lowerMessage.contains("平均") || 
            lowerMessage.contains("最大") || lowerMessage.contains("最小")) {
            
            StringBuilder response = new StringBuilder();
            response.append("我已经将您的查询转换为SQL语句：\n\n");
            response.append("**语义SQL结构：**\n");
            response.append("- 涉及表: ").append(
                semanticSQL.getTables() != null && !semanticSQL.getTables().isEmpty() 
                    ? String.join(", ", semanticSQL.getTables()) 
                    : "未指定"
            ).append("\n");
            response.append("- 查询列: ").append(
                semanticSQL.getColumns() != null && !semanticSQL.getColumns().isEmpty() 
                    ? String.join(", ", semanticSQL.getColumns()) 
                    : "所有列"
            ).append("\n");
            
            if (semanticSQL.getConditions() != null && !semanticSQL.getConditions().isEmpty()) {
                response.append("- 筛选条件: ").append(semanticSQL.getConditions().size()).append("个条件\n");
            }
            
            if (semanticSQL.getAggregations() != null && !semanticSQL.getAggregations().isEmpty()) {
                response.append("- 聚合函数: ").append(semanticSQL.getAggregations().size()).append("个\n");
            }
            
            if (semanticSQL.getJoins() != null && !semanticSQL.getJoins().isEmpty()) {
                response.append("- 表连接: ").append(semanticSQL.getJoins().size()).append("个\n");
            }
            
            response.append("\n**生成的MySQL SQL：**\n```sql\n").append(mysqlSQL).append("\n```\n\n");
            response.append("您想要执行这个SQL查询吗？");
            
            return response.toString();
        } else {
            return "我理解您想要：" + userMessage + "\n\n" +
                   "我已经生成了相应的SQL查询语句。您需要我执行查询并返回结果吗？";
        }
    }
}